matching-core/src/main/java/com/graphhopper/matching/MapMatching.java

```diff
@@ -269,26 +277,46 @@ public MatchResult doWork(List<GPXEntry> gpxList) {
     }
 
     /**
-     * Find the possible locations of each qpxEntry in the graph.
+     * Find the possible locations (edges) of each GPXEntry in the graph.
      */
-    private List<List<QueryResult>> lookupGPXEntries(List<GPXEntry> gpxList,
+    private List<Collection<QueryResult>> lookupGPXEntries(List<GPXEntry> gpxList,
                                                      EdgeFilter edgeFilter) {
 
-        List<List<QueryResult>> gpxEntryLocations = new ArrayList<>();
+        final List<Collection<QueryResult>> gpxEntryLocations = new ArrayList<>();
         for (GPXEntry gpxEntry : gpxList) {
-            gpxEntryLocations.add(locationIndex.findNClosest(gpxEntry.lat, gpxEntry.lon, edgeFilter,
-                    measurementErrorSigma));
+            final List<QueryResult> queryResults = locationIndex.findNClosest(
+                    gpxEntry.lat, gpxEntry.lon, edgeFilter, measurementErrorSigma);
+            gpxEntryLocations.add(queryResults);
         }
         return gpxEntryLocations;
     }
 
+    private List<Collection<QueryResult>> dedupeQueryResultsByClosestNode(
+            List<Collection<QueryResult>> queriesPerEntry) {
+        final List<Collection<QueryResult>> result = new ArrayList<>();
+
+        final Comparator<QueryResult> dedupeComparator = new Comparator<QueryResult>() {
```

Comment:
Instead of comparator and TreeSet can we use a `HashMap<closest node, query result>`? (less memory intense as array based)